File: XMLParsingTest/app/src/main/java/com/dstrube/xmlparsingtest/controller/MenuItemXMLHandler.java
package com.dstrube.xmlparsingtest.controller;

import java.util.ArrayList;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import com.dstrube.xmlparsingtest.model.MenuItem;

public class MenuItemXMLHandler extends DefaultHandler {

	private Boolean currentElement = false;
	private String currentValue = "";
	private MenuItem item = null;
	private ArrayList<MenuItem> list = new ArrayList<MenuItem>();

	public ArrayList<MenuItem> getList(){
		return list;
	}
	
	//When tag starts
	@Override
	public void startElement(String uri, String localName, String qName,
			Attributes attributes) throws SAXException {
		//super.startElement(uri, localName, qName, attributes);
		currentElement = true;
		currentValue = "";
		if (localName.equals("item")){
			item = new MenuItem();
		}
	}
	
	//When tag ends
	@Override
	public void endElement(String uri, String localName, String qName)
			throws SAXException {
		//super.endElement(uri, localName, qName);
		currentElement = false;
		if (localName.equalsIgnoreCase("id")){
			item.setId(currentValue);
		}
		else if (localName.equalsIgnoreCase("name")){
			item.setName(currentValue);
		}
		else if (localName.equalsIgnoreCase("cost")){
			item.setCost(currentValue);
		}
		else if (localName.equalsIgnoreCase("description")){
			item.setDescription(currentValue);
		}
		else if (localName.equalsIgnoreCase("item")){
            list.add(item);
		}
		//else?
	}
	
	//To get tag characters
	@Override
	public void characters(char[] ch, int start, int length)
			throws SAXException {
		//super.characters(ch, start, length);
		if (currentElement){
			currentValue = currentValue + new String(ch, start, length);
		}
	}
}

File: kotlin/fix/XMLParsingTest/app/src/main/java/com/dstrube/xmlparsingtest/controller/MenuItemXMLHandler.kt
package com.dstrube.xmlparsingtest.controller

import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler
import org.xml.sax.SAXException

import com.dstrube.xmlparsingtest.model.MenuItem

class MenuItemXMLHandler : DefaultHandler(){
    private var currentElement = false
    private var currentValue = ""
    private var item: MenuItem? = null
    private val list = ArrayList<MenuItem?>()


    fun getList(): ArrayList<MenuItem?> {
        return list
    }

    //When tag starts
    @Throws(SAXException::class)
    override fun startElement(uri: String, localName: String, qName: String,
                     attributes: Attributes) {
        //super.startElement(uri, localName, qName, attributes);
        currentElement = true
        currentValue = ""
        if (localName == "item") {
            item = MenuItem()
        }
    }

    //When tag ends
    @Throws(SAXException::class)
    override fun endElement(uri: String?, localName: String?, qName: String?) {
        //super.endElement(uri, localName, qName);
        currentElement = false
        if (localName!!.equals("id", ignoreCase = true)) {
            item?.setId(currentValue)
        } else if (localName.equals("name", ignoreCase = true)) {
            item?.setName(currentValue)
        } else if (localName.equals("cost", ignoreCase = true)) {
            item?.setCost(currentValue)
        } else if (localName.equals("description", ignoreCase = true)) {
            item?.setDescription(currentValue)
        } else if (localName.equals("item", ignoreCase = true)) {
            list.add(item)
        }
        //else?
    }

    //To get tag characters
    @Throws(SAXException::class)
    override fun characters(ch: CharArray?, start: Int, length: Int) {
        //super.characters(ch, start, length);
        if (currentElement) {
            currentValue += String(ch!!, start, length)
        }
    }
}